Start preparing completion XBlock for production

pmitros opened the pull request:
Wendy would like to DoneXBlock become production-ready. Product has also agreed it makes sense to bring DoneXBlock to production. This is a first step in that direction. This PR will:

* Move it from edx-private.txt into github.txt
* Add fixes for Studio
* Add fixes for eventing
* Add cleanup for edx-platform CSS issues

pmitros commented:
Code changes are: https://github.com/pmitros/DoneXBlock/pull/5

pmitros commented:
@explorerleslie @cptvitamin This is a small PR updating edx-platform to latest recommender. It:

* Adds Studio support to the completion XBlock, fixes a minor CSS styling issue. 
* Fixes a CSS styling issue
* Deploys previous accessibility fixes (which I realized, making this, were in the DoneXBlock repo, but not on edx.org, which has an old revision)
* General code cleanup.

Diff from the last revision: 
https://github.com/pmitros/DoneXBlock/pull/5

Last revision was reviewed by @singingwolfboy. I'm not sure if this should be reviewed by T&L, David, or someone else. 

It's a very small piece of code, so review is likely to be very fast.

pmitros commented:
@explorerleslie Just to keep you in the loop: Mark did a very nice accessibility review. I *believe* I just addressed all of the issues (although he may want to confirm that I addressed them correctly).

explorerleslie commented:
@pmitros I've created https://openedx.atlassian.net/browse/TNL-3599 for this PR's TNL review. Can you please update the screenshots in the readme? I know they've undergone a revision that's not reflected here.

explorerleslie commented:
@pmitros also, one comment on the button text, which we've talked about before. One thought is that, the "completed" state could just say "Completed". Because the button acts like a toggle, I think calling out the undo state might be unnecessary. What do you think?

pmitros commented:
@cptvitamin We just discussed this in the hallway. Would you mind reconfirming the final version is correctly accessible? It should behave just like a checkbox.

cahrens commented:
@pmitros please rebase this branch on master (note that the django 1.8 upgrade is now in master) and include a short description of how to enable and test this xblock (link a sandbox if there is any substantial setup). We plan to review this PR within the next 2 weeks.

Also, what is the status of the accessibility review?

@explorerleslie If this is "production-ready", does that mean it will be promoted out of advanced_modules? If so, I would expect we want automated test coverage for it at that point.

pmitros commented:
@cahrens Thank you. I rebased to master/Django 1.8. As per your suggestion, I added tests as well.

In this case, supported does not mean promoting out of `advanced_modules` (yet), but adding documentation for course staffs on docs.edx.org. However, there will be a second PR which will bring many of the advanced modules as enabled by default, and this will be among them. We have a lot of very nice functionality on edX which course authors are not using, and we would like to make it more visible.

cahrens commented:
jenkins run all

cahrens commented on openedx/tests/xblock_integration/test_done.py:
```diff
@@ -0,0 +1,139 @@
+        # Nose runs setUpClass methods even if a class decorator says to skip
+        # the class: https://github.com/nose-devs/nose/issues/946
+        # So, skip the test class here if we are not in the LMS.
+        if settings.ROOT_URLCONF != 'lms.urls':
```
You can write this with a decorator--

@unittest.skipUnless(settings.ROOT_URLCONF == 'lms.urls', 'Test only valid in lms')

pmitros replied:
I believe that's correct.

cahrens commented on openedx/tests/xblock_integration/test_done.py:
```diff
@@ -0,0 +1,139 @@
+            cls.chapter = ItemFactory.create(
+                parent=cls.course, display_name='Overview'
+            )
+            cls.section = ItemFactory.create(
```
If category is not specified, I believe it defaults to "chapter". Therefore, if you want the items being created to match your intended structure, you should specify the category for section and unit.

pmitros replied:
Specified now. Thank you.

cahrens commented on openedx/tests/xblock_integration/test_done.py:
```diff
@@ -0,0 +1,139 @@
+        super(TestDone, self).setUp()
+        for idx, student in enumerate(self.STUDENTS):
+            username = "u{}".format(idx)
+            self.create_account(username, student['email'], student['password'])
```
You can use UserFactory.

pmitros replied:
I'm okay either way. Eventually, we will need to decouple from LMS, but that's a ways in the future. This code is actually a little bit shorter than with UserFactory. Let me know if you want to change.

cahrens replied:
I would leave it out because it introduces a dependency in openedx on LMS (in reality, that code should be moved from LMS to a more common location).

cahrens commented on openedx/tests/xblock_integration/test_done.py:
```diff
@@ -0,0 +1,139 @@
+        Get url for the specified xblock handler
+        """
+        return reverse('xblock_handler', kwargs={
+            'course_id': self.course.id.to_deprecated_string(),
```
Instead of using to_deprecated_string, just do unicode(self.course.id)

pmitros replied:
I fixed this globally. Thank you.

cahrens commented on openedx/tests/xblock_integration/test_done.py:
```diff
@@ -0,0 +1,139 @@
+        """
+        return reverse('xblock_handler', kwargs={
+            'course_id': self.course.id.to_deprecated_string(),
+            'usage_id': quote_slashes(self.course.id.make_usage_key('done', xblock_name).to_deprecated_string()),
```
Same thing here, use unicode instead of to_deprecated_string (also don't think you need quote_slashes).

unicode(self.course.id.make_usage_key('done', xblock_name))

pmitros replied:
quote_slashes was, indeed, unnecessary as well. I don't completely understand the locators, so this was copied verbatim from other tests. Thank you.

cahrens commented on openedx/tests/xblock_integration/test_done.py:
```diff
@@ -0,0 +1,139 @@
+    of the block, make sure the page renders, toggle them a few times,
+    make sure they've toggled, and reconfirm the page renders.
+    """
+    STUDENTS = [
```
Why do you have 2 students? You only use the first one.

pmitros replied:
We're at one. Thank you. Missed that when cutting-and-pasting from another test.

benpatterson commented:
jenkins run all

cahrens commented:
@pmitros The bok choy failure is a known issue (https://openedx.atlassian.net/browse/TNL-3853).

cahrens commented:
@pmitros I'm done with my first review. Thank you for the Python unit test covering the server side. Note though that there is still no test that verifies the JS works properly (and thus the server side is ever called). I'm going to check in with @jbarciauskas about how we want to handle testing of supported xblocks-- I'll cc you on the conversation.

pmitros commented:
@mulby I made eventing changes as discussed: 

* Switched to `runtime.publish`. Confirmed `runtime.publish` works.
* Event tests omitted, as per conversation. Encouraging adding tests to the platform for `runtime.publish` (which is currently the recommended way to generate events in XBlocks, but has no tests).
* Continuing to use same event name for client-side and server-side
* However, switching event names to past tense.

pmitros commented:
@cahrens Are we good to merge once I bump requirements.txt to latest, rebase, and tests pass? I believe all issues which are addressable have been addressed. Outstanding platform issues identified: 

* `runtime.publish` in the platform identified to not have tests.
* XBlocks needs a test framework for JS/visual diff/event testing. @benpatterson is working on stories towards this. I made a PR proposing such a test framework (https://github.com/edx/edx-platform/pull/10831) but we need to go through arch review and flush it out.
* XBlocks needs full i18n.
* We need some sanity in naming XBlocks. We have several different names used in different places in the platform (esp. in Studio)

mulby commented:
@pmitros I didn't realize we were talking about test cases in edx-platform. I thought you were talking about unit tests in your block repo.

I would prefer you include at least one test here that verifies that when it's toggled an event is emitted, which will ensure that if someone up-revs the block repo, the "contract" with the platform is maintained. I would expect this test to fail if a future developer put some conditional logic around the event emission statement or removed it entirely. I do agree with you that is probably an unlikely scenario, however, I also see the adding of the test as a very low effort addition to this PR that helps clarify the contract.

pmitros commented:
@mulby Roger wilco.

pmitros commented:
@cahrens Please hold off until said test is in place

pmitros commented:
@cahrens Event testing is in place. I think this is good to merge?

benpatterson commented:
We don't have event-testing-related methods elsewhere (outside of bok-choy)? @mulby ?

cahrens commented on openedx/tests/xblock_integration/test_done.py:
```diff
@@ -0,0 +1,233 @@
+from lms.djangoapps.courseware.tests.factories import GlobalStaffFactory
+
+
+class SimpleEventTestMixin(object):
```
We definitely want to move this into a location where it can be shared. Can you at least put it in a util file in openedx/tests/xblock_integration?

I see that it is similar to EventTestMixin in common/djangoapps/util/testing.py. Would it be possible to just use EventTestMixin, supplying your own tracker? Or is there not enough shared code to make it worth it?

mulby replied:
I would strongly prefer this is merged with EventTestMixin. The core of the problem is probably that there are still two different methods used to emit events (eventtracking and log_event), but we should just extend EventTestMixin to support both.

mulby replied:
Also, you should be able to leverage this:

https://github.com/edx/edx-platform/blob/master/openedx/core/lib/tests/assertions/events.py

Which provides most of the flexible assertion magic that is available in the acceptance tests.

pmitros replied:
I definitely plan on doing this, but I would prefer to do so in a week or two, either with the XBlock testing PR or the RateXBlock PR. I am talking to @benpatterson about how we want to structure XBlocks tests, and we'll also have a better idea of whether this is properly generic once it is working on two XBlocks. Before then, I'd rather not make this a common utility. 

If I need it for the :+1:, I'm okay doing it now, but I think that'd be a more logical time to do so.

I started with EventTestMixin, but it didn't really contribute much. It just added a thin layer of indirection between itself and a tracker (where most of the functionality lived). It added a fair bit of complexity, wasn't really documented, and didn't save on any code. The examples of how it was used were a little bit messy. Ben suggested looking at the cohorts one, which was a lot more developed and fairly nice (it puts events in Mongo, and you search from them), but Ben said that it requires bok_choy. When I played with moving this into bok_choy, testing got very slow.

(The docstring was intended to convey this)

cahrens commented on openedx/tests/xblock_integration/test_done.py:
```diff
@@ -0,0 +1,233 @@
+        Create users
+        """
+        super(TestDone, self).setUp()
+        #EventTestMixin.setUp(self, mock.Mock())
```
Remove commented out line.

cahrens commented:
A nit about a commented out line, and I would like your testing mixin moved into a utility file/folder so it can easily be shared by other tests (extending the existing EventTestMixin if that makes sense). Otherwise :+1: (and I don't need to look at this again).

Let me go give the other related PR one last look....

cahrens commented on openedx/tests/xblock_integration/test_done.py:
```diff
@@ -0,0 +1,233 @@
+        self.assert_event_emitted('edx.done.toggled', event_fields={"done": True})
+        self.reset_tracker()
+        self.check_ajax(1, {'done': False}, False)
+        self.assert_event_emitted('edx.done.toggled', event_fields={"done": False})
```
Looking at https://github.com/pmitros/DoneXBlock/pull/5/files, you are also now emitting a "grade" event. You should test for it as well.

pmitros replied:
I added a test for this. Note that this is a new Mixin. Grading events are not analytics events, and go through a different pipeline which is why this is a little more code than might be expected (see https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/module_render.py#L520 for the logic around this).

pmitros commented:
I believe all comments were either implemented or addressed. I rebased, and tests pass locally.

pmitros commented:
Oh, except mulby's new comments. I'll talk to him in person.

pmitros commented:
@mulby I renamed the Mixin to be called DoneEventTestMixin. Putting them in the same giant class makes things a lot more ugly. Renaming also makes it clear this is not a design pattern. I also added much more commentary to the top saying "Please don't use this elsewhere yet."

mulby commented:
@pmitros thanks for doing that :+1: 

I would prefer you mock at runtime.publish instead of log_event, but won't hold up the PR on that change.

pmitros commented:
jenkins run all

pmitros commented:
@mulby Thanks. I agree that would be a better mock point. I'd rather clear this up in the next few PRs.

jzoldak commented:
jenkins run all

jzoldak commented:
@pmitros I'm not sure what is going on with this PR and jenkins. I'm going to try closing it and reopening because I suspect it may have something to do with the cache of one of the jenkins plug-ins (ghprb).

jzoldak commented:
@pmitros yeah, it looks like that kicked it. The tests are running now.

pmitros commented:
jenkins run all

jzoldak commented:
jenkins run bokchoy
